stl/src/atomic_wait.cpp

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// implement atomic wait / notify_one / notify_all
+
+#include <atomic>
+#include <cstdint>
+#include <new>
+#include <thread>
+
+#include <Windows.h>
+
+namespace {
+
+    constexpr std::size_t _Wait_table_size_power = 8;
+    constexpr std::size_t _Wait_table_size       = 1 << _Wait_table_size_power;
+    constexpr std::size_t _Wait_table_index_mask = _Wait_table_size - 1;
+
+#pragma warning(push)
+#pragma warning(disable : 4324) // structure was padded due to alignment specifier
+    struct alignas(std::hardware_destructive_interference_size) _Wait_table_entry {
+        // Arbitraty variable to wait/notify on if target wariable is not proper atomic for that
+        // Size is largest of lock-free to make aliasing problem into hypothetical
+        std::atomic<unsigned long long> _Counter;
+
+        CONDITION_VARIABLE _Condition = CONDITION_VARIABLE_INIT;
+        SRWLOCK _Lock                 = SRWLOCK_INIT;
+    };
+#pragma warning(pop)
+
+    _Wait_table_entry& _Atomic_wait_table_entry(const void* const _Storage) noexcept {
+        static _Wait_table_entry wait_table[_Wait_table_size];
+        auto index = reinterpret_cast<std::uintptr_t>(_Storage);
+        index ^= index >> (_Wait_table_size_power * 2);
+        index ^= index >> _Wait_table_size_power;
+        return wait_table[index & _Wait_table_index_mask];
+    }
+
+    static constexpr std::size_t _Uninitialized_spin_count = (std::numeric_limits<std::size_t>::max)();
+    static std::atomic<std::size_t> _Atomic_spin_count     = _Uninitialized_spin_count;
+
+    std::size_t _Atomic_init_spin_count() noexcept {
+        std::size_t result = (std::thread::hardware_concurrency() == 1 ? 0 : 10'000) * _Atomic_spin_value_step;
+        _Atomic_spin_count.store(result, std::memory_order_relaxed);
+        // Make sure other thread is likely to get this,
+        // as we've done kernel call for that.
+        std::atomic_thread_fence(std::memory_order_seq_cst);
+        return result;
+    }
+
+
+#if _STL_WIN32_WINNT >= _WIN32_WINNT_WIN8
+    constexpr bool _Have_wait_functions() {
+        return true;
+    }
+#define __crtWaitOnAddress       WaitOnAddress
+#define __crtWakeByAddressSingle WakeByAddressSingle
+#define __crtWakeByAddressAll    WakeByAddressAll
+
+#pragma comment(lib, "Synchronization.lib")
+
+    void _Atomic_wait_fallback(

```

Comment:
These should be `[[noreturn]]` so that the compiler knows they're unlikely.

Reply:
Ok, I'll put `[[noreturn]]`
They are in a wrong if branch that is controlled by a constant condition. So it should not matter much. Compiler will eliminate them anyway.